rust/parquet_derive/src/parquet_field.rs

```diff
@@ -0,0 +1,942 @@
+                    Type::TypePath(_) => self.option_into_vals(),
+                    _ => unimplemented!("sorry charlie"),
+                },
+                ref f @ _ => unimplemented!("whoa: {:#?}", f),
+            },
+            Type::Reference(_, ref first_type) => match **first_type {
+                Type::TypePath(_) => self.copied_direct_vals(),
+                Type::Option(ref second_type) => match **second_type {
+                    Type::TypePath(_) => self.option_into_vals(),
+                    Type::Reference(_, ref second_type) => match **second_type {
+                        Type::TypePath(_) => self.option_into_vals(),
+                        _ => unimplemented!("sorry charlie"),
+                    },
+                    ref f @ _ => unimplemented!("whoa: {:#?}", f),
+                },
+                ref f @ _ => unimplemented!("whoa: {:#?}", f),
+            },
+            f @ _ => unimplemented!("don't support {:#?}", f),
+        };
+
+        let definition_levels = match &self.ty {
+            Type::TypePath(_) => None,
+            Type::Option(ref first_type) => match **first_type {
+                Type::TypePath(_) => Some(self.optional_definition_levels()),
+                Type::Option(_) => unimplemented!("nested options? that's weird"),
+                Type::Reference(_, ref second_type)
+                | Type::Vec(ref second_type)
+                | Type::Array(ref second_type) => match **second_type {
+                    Type::TypePath(_) => Some(self.optional_definition_levels()),
+                    _ => unimplemented!("a little too much nesting. bailing out."),
+                },
+            },
+            Type::Reference(_, ref first_type)
+            | Type::Vec(ref first_type)
+            | Type::Array(ref first_type) => match **first_type {
+                Type::TypePath(_) => None,
+                Type::Reference(_, ref second_type)
+                | Type::Vec(ref second_type)
+                | Type::Array(ref second_type)
+                | Type::Option(ref second_type) => match **second_type {
+                    Type::TypePath(_) => Some(self.optional_definition_levels()),
+                    Type::Reference(_, ref third_type) => match **third_type {
+                        Type::TypePath(_) => Some(self.optional_definition_levels()),
+                        _ => unimplemented!(
+                            "we don't do some more complex definition levels... yet!"
+                        ),
+                    },
+                    _ => unimplemented!(
+                        "we don't do more complex definition levels... yet!"
+                    ),
+                },
+            },
+        };
+
+        let write_batch_expr = if definition_levels.is_some() {
+            quote! {
+                if let #column_writer(ref mut typed) = column_writer {
+                    typed.write_batch(&vals[..], Some(&definition_levels[..]), None).unwrap();
```

Comment:
It would be good if we can add comment to point out where the `vals` and `definition_levels` are from, to make this easier to maintain.